src/EventListener/UserSessionListener.php

```diff
@@ -0,0 +1,274 @@
+     * on the container scope.
+     *
+     * @param FilterResponseEvent $event The event object
+     */
+    public function onKernelResponse(FilterResponseEvent $event)
+    {
+        if (!$this->isFrontendMasterRequest($event)
+            && !$this->isBackendMasterRequest($event)
+        ) {
+            return;
+        }
+
+        $request = $event->getRequest();
+
+        if ($this->isBackendScope()) {
+            $this->storeBackendSession($request);
+            return;
+        }
+
+        $this->storeFrontendSession($request);
+    }
+
+    /**
+     * Stores the session for the back end.
+     *
+     * @param Request $request
+     */
+    private function storeBackendSession(Request $request)
+    {
+        if (!$this->checkShouldNotModifyBackendSession($request)) {
+            $this->storeSession();
+            return;
+        }
+
+        $key        = $request->query->has('popup') ? 'popupReferer' : 'referer';
+        $refererId  = $request->attributes->get('_contao_referer_id');
+        /** @var AttributeBagInterface $bag */
+        $bag        = $this->getSessionBag();
+        $refererOld = $bag->get($key);
+        $refererNew = [];
+
+        if (!is_array($refererOld)
+            || !is_array($refererOld[$refererId])
+        ) {
+            $refererOld[$refererId]['last'] = '';
+        }
+
+        while (count($refererOld) >= 25) {
+            array_shift($refererOld);
+        }
+
+        $ref = $request->query->get('ref');
+
+        if ($ref != '' && isset($refererOld[$ref])) {
+            if (!isset($refererOld[$refererId])) {
+                $refererOld[$refererId] = [];
+            }
+
+            $refererNew[$refererId]         = array_merge(
+                $refererOld[$refererId],
+                $refererOld[$ref]
+            );
+            $refererNew[$refererId]['last'] = $refererOld[$ref]['current'];
+        } elseif (count($refererOld) > 1) {
+            $refererNew[$refererId] = end($refererOld);
+        }
+
+        $refererNew[$refererId]['current'] = $request->getRequestUri();
+
+        $bag->set($key, $refererNew);
+
+        $this->storeSession();
+    }
+
+    /**
+     * Check if we should not modify the session in the back end.
+     *
+     * @param Request $request
+     *
+     * @return bool
+     */
+    private function checkShouldNotModifyBackendSession(Request $request)
+    {
+        if (!$request->query->has('act')
+            && !$request->query->has('key')
+            && !$request->query->has('token')
+            && !$request->query->has('state')
+            && 'feRedirect' !== $request->query->get('do')
+            && !$request->isXmlHttpRequest()
+        ) {
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Stores the session for the front end.
+     *
+     * @param Request $request
+     */
+    private function storeFrontendSession(Request $request)
+    {
+        /** @var AttributeBagInterface $bag */
```

Comment:
Ah, that was needed before I put this in its own method. Removed in 4bb2e5c952f443529c76332796389cb12ccaf011.